src/main/java/com/epam/aidial/core/storage/Ec2InstanceMetadataCredentialProvider.java

```diff
@@ -0,0 +1,119 @@
+package com.epam.aidial.core.storage;
+
+import com.epam.aidial.core.util.ProxyUtil;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.jclouds.aws.domain.SessionCredentials;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+@Slf4j
+public class Ec2InstanceMetadataCredentialProvider implements CredentialProvider {
+
+    private static final String EC2_INSTANCE_METADATA_BASE_URL = "http://169.254.169.254/latest/";
+    private static final String EC2_INSTANCE_METADATA_CREDENTIALS_URL = EC2_INSTANCE_METADATA_BASE_URL + "meta-data/iam/security-credentials/";
+    private static final String EC2_TOKEN_TTL_HEADER_NAME = "X-aws-ec2-metadata-token-ttl-seconds";
+    private static final String EC2_METADATA_TOKEN_HEADER_NAME = "X-aws-ec2-metadata-token";
+
+    private final HttpClient httpClient;
+    private final Lock lock = new ReentrantLock();
```

Comment:
I believe there is no benefit using a reentrant lock here. It could be replaced with `synchronized` or AtomicReference.

Reply:
fixed